SendJSON crashes with nil pointer when called from SendError path

## Description

`SendJSON` crashes with a nil pointer dereference when JSON serialization fails during error handling. The crash occurs because `SendError` intentionally passes `nil` for the request parameter when dispatching to `SendJSONError`, which then calls `SendJSON(w, nil, err)`. However, `SendJSON` unconditionally calls `r.Context()` for logging, causing a panic.

## Stack Trace

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x104d67914]

goroutine 83 [running]:
github.com/go-fuego/fuego.init.func5(0x14000582000, 0x0, {0x1051645e0, 0x140006e4480})
	fuego@v0.19.0/serialization.go:169 +0x264
github.com/go-fuego/fuego.SendJSONError(...)
	fuego@v0.19.0/serialization.go:217
github.com/go-fuego/fuego.SendError(0x14000582000, 0x0, {0x1051645e0, 0x140006e4480})
	fuego@v0.19.0/serialization.go:194 +0x1a4
```

## Root Cause

In commit `8470afc` ("BREAKING: use slog.\*\*\*Context where applicable"), `slog.Error` was changed to `slog.ErrorContext(r.Context(), ...)` in `SendJSON`:

```go
// serialization.go line 169
slog.ErrorContext(r.Context(), "cannot serialize JSON", "error", err, "ans", ans)
```

However, `SendError` explicitly passes `nil` for the request when calling type-specific error handlers:

```go
// serialization.go lines 191-194
case "application/json", "":
    return SendJSONError(w, nil, err)  // nil request!
```

And `SendJSONError` ignores the request parameter entirely:

```go
// serialization.go line 214
var SendJSONError ErrorSender = func(w http.ResponseWriter, _ *http.Request, err error) error {
    // ...
    return SendJSON(w, nil, httpError)  // passes nil to SendJSON
}
```

## Reproduction

1. Return any serializable struct from a fuego controller
2. Have that struct fail JSON marshaling (e.g., contains a channel, func, or causes an encoding error)
3. Fuego calls `SendError` → `SendJSONError` → `SendJSON` with nil request
4. `SendJSON` tries to log with `slog.ErrorContext(r.Context(), ...)` → **panic**

## Suggested Fix

Add a nil check before using `r.Context()` in `SendJSON`:

```go
if err != nil {
    if r != nil {
        slog.ErrorContext(r.Context(), "cannot serialize JSON", "error", err, "ans", ans)
    } else {
        slog.Error("cannot serialize JSON", "error", err, "ans", ans)
    }
    return err
}
```

The same fix should be applied to `SendYAML`, `SendXML`, `SendHTML`, and `SendText`.

Alternatively, `SendError` and the `Send*Error` functions could be updated to properly pass through the request parameter.

## Affected Versions

- **Introduced in**: Commits after v0.18.0 (specifically commit `8470afc` on April 23, 2025)
- **Last working version**: v0.18.0

## Workaround

Override `fuego.SendJSON` with a nil-safe implementation:

```go
func init() {
    fuego.SendJSON = func(w http.ResponseWriter, r *http.Request, ans any) error {
        w.Header().Set("Content-Type", "application/json")
        encoder := json.NewEncoder(w)
        err := encoder.Encode(ans)
        if err != nil {
            if r != nil {
                slog.ErrorContext(r.Context(), "cannot serialize JSON", "error", err)
            } else {
                slog.Error("cannot serialize JSON", "error", err)
            }
            return err
        }
        return nil
    }
}
```

--- a/middleware/basicauth/basicauth.go
+++ b/middleware/basicauth/basicauth.go
@@ -42,7 +42,7 @@ func New(config Config) func(http.Handler) http.Handler {
 			}
 
 			w.Header().Set("WWW-Authenticate", `Basic realm="Restricted"`)
-			fuego.SendJSONError(w, nil, err)
+			fuego.SendJSONError(w, r, err)
 		})
 	}
 }

--- a/security.go
+++ b/security.go
@@ -230,7 +230,7 @@ func (security Security) TokenToContext(searchFunc ...func(*http.Request) string
 			// Validate the token
 			t, err := security.ValidateToken(token)
 			if err != nil {
-				SendJSONError(w, nil, err)
+				SendJSONError(w, r, err)
 				return
 			}
 
@@ -318,20 +318,20 @@ func authWall(authorizeFunc func(userRoles ...string) bool) func(next http.Handl
 			// Get the authorizationHeader from the context (set by TokenToContext)
 			claims, err := TokenFromContext(r.Context())
 			if err != nil {
-				SendJSONError(w, nil, UnauthorizedError{Title: "Unauthorized"})
+				SendJSONError(w, r, UnauthorizedError{Title: "Unauthorized"})
 				return
 			}
 
 			// Get the subject and userRoles from the claims
 			userRoles, ok := claims.(jwt.MapClaims)["roles"].([]string)
 			if !ok {
-				SendJSONError(w, nil, UnauthorizedError{Title: "Could not find roles in token"})
+				SendJSONError(w, r, UnauthorizedError{Title: "Could not find roles in token"})
 				return
 			}
 
 			// Check if the user is authorized
 			if !authorizeFunc(userRoles...) {
-				SendJSONError(w, nil, ForbiddenError{Title: "Access denied"})
+				SendJSONError(w, r, ForbiddenError{Title: "Access denied"})
 				return
 			}
 
@@ -381,14 +381,14 @@ func (security Security) StdLoginHandler(verifyUserInfo func(r *http.Request) (j
 	return func(w http.ResponseWriter, r *http.Request) {
 		claims, err := verifyUserInfo(r)
 		if err != nil {
-			SendJSONError(w, nil, err)
+			SendJSONError(w, r, err)
 			return
 		}
 
 		// Send the token to the cookies
 		token, err := security.GenerateTokenToCookies(claims, w)
 		if err != nil {
-			SendJSONError(w, nil, err)
+			SendJSONError(w, r, err)
 			return
 		}
 
@@ -476,14 +476,14 @@ func (security Security) LoginHandler(verifyUserInfo func(user, password string)
 func (security Security) RefreshHandler(w http.ResponseWriter, r *http.Request) {
 	claims, err := TokenFromContext(r.Context())
 	if err != nil {
-		SendJSONError(w, nil, UnauthorizedError{Title: "Could not find token in context"})
+		SendJSONError(w, r, UnauthorizedError{Title: "Could not find token in context"})
 		return
 	}
 
 	// Send the token to the cookies
 	token, err := security.GenerateTokenToCookies(claims, w)
 	if err != nil {
-		SendJSONError(w, nil, err)
+		SendJSONError(w, r, err)
 		return
 	}
 

--- a/serialization.go
+++ b/serialization.go
@@ -143,7 +143,7 @@ var SendYAML = func(w http.ResponseWriter, r *http.Request, ans any) (err error)
 
 // SendYAMLError sends a YAML error response.
 // If the error implements ErrorWithStatus, the status code will be set.
-func SendYAMLError(w http.ResponseWriter, _ *http.Request, err error) {
+func SendYAMLError(w http.ResponseWriter, r *http.Request, err error) {
 	status := http.StatusInternalServerError
 	var errorStatus ErrorWithStatus
 	if errors.As(err, &errorStatus) {
@@ -151,7 +151,7 @@ func SendYAMLError(w http.ResponseWriter, _ *http.Request, err error) {
 	}
 
 	w.WriteHeader(status)
-	_ = SendYAML(w, nil, err)
+	_ = SendYAML(w, r, err)
 }
 
 // SendJSON sends a JSON response.
@@ -181,15 +181,15 @@ var SendError = func(w http.ResponseWriter, r *http.Request, err error) {
 	for _, header := range parseAcceptHeader(r.Header) {
 		switch inferAcceptHeader(header, nil) {
 		case "application/xml":
-			SendXMLError(w, nil, err)
+			SendXMLError(w, r, err)
 		case "text/html":
-			SendHTMLError(w, nil, err)
+			SendHTMLError(w, r, err)
 		case "text/plain":
 			SendTextError(w, r, err)
 		case "application/json":
-			SendJSONError(w, nil, err)
+			SendJSONError(w, r, err)
 		case "application/x-yaml", "text/yaml; charset=utf-8", "application/yaml": // https://www.rfc-editor.org/rfc/rfc9512.html
-			SendYAMLError(w, nil, err)
+			SendYAMLError(w, r, err)
 		default:
 			continue
 		}
@@ -200,7 +200,7 @@ var SendError = func(w http.ResponseWriter, r *http.Request, err error) {
 
 // SendJSONError sends a JSON error response.
 // If the error implements ErrorWithStatus, the status code will be set.
-func SendJSONError(w http.ResponseWriter, _ *http.Request, err error) {
+func SendJSONError(w http.ResponseWriter, r *http.Request, err error) {
 	status := http.StatusInternalServerError
 	var errorStatus ErrorWithStatus
 	if errors.As(err, &errorStatus) {
@@ -215,7 +215,7 @@ func SendJSONError(w http.ResponseWriter, _ *http.Request, err error) {
 	}
 
 	w.WriteHeader(status)
-	_ = SendJSON(w, nil, err)
+	_ = SendJSON(w, r, err)
 }
 
 // SendXML sends a XML response.
@@ -248,7 +248,7 @@ func SendXMLError(w http.ResponseWriter, r *http.Request, err error) {
 	}
 
 	w.WriteHeader(status)
-	err = SendXML(w, nil, err)
+	err = SendXML(w, r, err)
 	if err != nil {
 		slog.ErrorContext(r.Context(), "Cannot serialize returned response to XML", "error", err)
 		_, _ = w.Write([]byte(`{"error":"Cannot serialize returned response to XML"}`))
@@ -294,7 +294,7 @@ var SendHTML = func(w http.ResponseWriter, r *http.Request, ans any) error {
 
 // SendHTMLError sends a HTML response.
 // If the error implements ErrorWithStatus, the status code will be set.
-func SendHTMLError(w http.ResponseWriter, _ *http.Request, err error) {
+func SendHTMLError(w http.ResponseWriter, r *http.Request, err error) {
 	status := http.StatusInternalServerError
 	var errorStatus ErrorWithStatus
 	if errors.As(err, &errorStatus) {
@@ -313,7 +313,7 @@ func SendHTMLError(w http.ResponseWriter, _ *http.Request, err error) {
 
 // SendText sends a HTML response.
 // Declared as a variable to be able to override it for clients that need to customize serialization.
-func SendText(w http.ResponseWriter, _ *http.Request, ans any) error {
+func SendText(w http.ResponseWriter, r *http.Request, ans any) error {
 	var err error
 	w.Header().Set("Content-Type", "text/plain; charset=utf-8")
 	stringToWrite, ok := ans.(string)
@@ -332,7 +332,7 @@ func SendText(w http.ResponseWriter, _ *http.Request, ans any) error {
 
 // SendTextError sends a Text response.
 // If the error implements ErrorWithStatus, the status code will be set.
-func SendTextError(w http.ResponseWriter, _ *http.Request, err error) {
+func SendTextError(w http.ResponseWriter, r *http.Request, err error) {
 	status := http.StatusInternalServerError
 	var errorStatus ErrorWithStatus
 	if errors.As(err, &errorStatus) {
@@ -343,10 +343,10 @@ func SendTextError(w http.ResponseWriter, _ *http.Request, err error) {
 	var httpError HTTPError
 	if errors.As(err, &httpError) {
 		httpError.Status = status
-		_ = SendText(w, nil, httpError.PublicError())
+		_ = SendText(w, r, httpError.PublicError())
 		return
 	}
-	_ = SendText(w, nil, err.Error())
+	_ = SendText(w, r, err.Error())
 }
 
 func InferAcceptHeaderFromType(ans any) string {

--- a/serialization_test.go
+++ b/serialization_test.go
@@ -233,7 +233,7 @@ func TestJSONError(t *testing.T) {
 	err := validate(me)
 	w := httptest.NewRecorder()
 	err = ErrorHandler(context.Background(), err)
-	SendJSONError(w, nil, err)
+	SendJSONError(w, httptest.NewRequest("", "/", nil), err)
 	require.Equal(t, "application/json", w.Header().Get("Content-Type"))
 
 	require.JSONEq(t, `
@@ -304,7 +304,7 @@ func TestJSONError(t *testing.T) {
 
 func TestSendText(t *testing.T) {
 	w := httptest.NewRecorder()
-	SendText(w, nil, "Hello World")
+	SendText(w, httptest.NewRequest("", "/", nil), "Hello World")
 
 	require.Equal(t, "Hello World", w.Body.String())
 }
@@ -312,7 +312,7 @@ func TestSendText(t *testing.T) {
 func TestSendTextError(t *testing.T) {
 	t.Run("base", func(t *testing.T) {
 		w := httptest.NewRecorder()
-		SendTextError(w, nil, errors.New("Hello World"))
+		SendTextError(w, httptest.NewRequest("", "/", nil), errors.New("Hello World"))
 
 		assert.Equal(t, http.StatusInternalServerError, w.Result().StatusCode)
 		assert.Equal(t, "text/plain; charset=utf-8", w.Header().Get("Content-Type"))
@@ -320,7 +320,7 @@ func TestSendTextError(t *testing.T) {
 	})
 	t.Run("error with status", func(t *testing.T) {
 		w := httptest.NewRecorder()
-		SendTextError(w, nil, BadRequestError{Err: errors.New("Hello World")})
+		SendTextError(w, httptest.NewRequest("", "/", nil), BadRequestError{Err: errors.New("Hello World")})
 		assert.Equal(t, http.StatusBadRequest, w.Result().StatusCode)
 		assert.Equal(t, "text/plain; charset=utf-8", w.Header().Get("Content-Type"))
 		assert.Equal(t, "400 Bad Request", w.Body.String())
@@ -360,7 +360,7 @@ func TestSendYAML(t *testing.T) {
 func TestSendYAMLError(t *testing.T) {
 	t.Run("base", func(t *testing.T) {
 		w := httptest.NewRecorder()
-		SendYAMLError(w, nil, errors.New("Hello World"))
+		SendYAMLError(w, httptest.NewRequest("", "/", nil), errors.New("Hello World"))
 
 		require.Equal(t, http.StatusInternalServerError, w.Result().StatusCode)
 		require.Equal(t, "application/x-yaml", w.Header().Get("Content-Type"))
@@ -375,14 +375,14 @@ func TestSendYAMLError(t *testing.T) {
 	})
 	t.Run("error with status and detail", func(t *testing.T) {
 		w := httptest.NewRecorder()
-		SendYAMLError(w, nil, BadRequestError{Err: errors.New("Hello World"), Detail: "World, Hello"})
+		SendYAMLError(w, httptest.NewRequest("", "/", nil), BadRequestError{Err: errors.New("Hello World"), Detail: "World, Hello"})
 		require.Equal(t, http.StatusBadRequest, w.Result().StatusCode)
 		require.Equal(t, "application/x-yaml", w.Header().Get("Content-Type"))
 		require.Equal(t, crlf(`detail: World, Hello`), w.Body.String())
 	})
 	t.Run("error with multiple fields", func(t *testing.T) {
 		w := httptest.NewRecorder()
-		SendYAMLError(w, nil, BadRequestError{
+		SendYAMLError(w, httptest.NewRequest("", "/", nil), BadRequestError{
 			Err:    errors.New("Hello World"),
 			Detail: "World, Hello",
 			Title:  "Error: Hello, World",
@@ -448,7 +448,7 @@ func TestSendHTML(t *testing.T) {
 
 	t.Run("Renderer", func(t *testing.T) {
 		w := httptest.NewRecorder()
-		err := SendHTML(w, nil, &testRenderer{})
+		err := SendHTML(w, httptest.NewRequest("", "/", nil), &testRenderer{})
 		require.NoError(t, err)
 		require.Equal(t, "text/html; charset=utf-8", w.Header().Get("Content-Type"))
 		require.Equal(t, "hello", w.Body.String())
@@ -456,7 +456,7 @@ func TestSendHTML(t *testing.T) {
 
 	t.Run("HTML", func(t *testing.T) {
 		w := httptest.NewRecorder()
-		err := SendHTML(w, nil, HTML("hello"))
+		err := SendHTML(w, httptest.NewRequest("", "/", nil), HTML("hello"))
 		require.NoError(t, err)
 		require.Equal(t, "text/html; charset=utf-8", w.Header().Get("Content-Type"))
 		require.Equal(t, "hello", w.Body.String())
@@ -464,7 +464,7 @@ func TestSendHTML(t *testing.T) {
 
 	t.Run("string", func(t *testing.T) {
 		w := httptest.NewRecorder()
-		err := SendHTML(w, nil, "hello")
+		err := SendHTML(w, httptest.NewRequest("", "/", nil), "hello")
 		require.NoError(t, err)
 		require.Equal(t, "text/html; charset=utf-8", w.Header().Get("Content-Type"))
 		require.Equal(t, "hello", w.Body.String())
@@ -472,7 +472,7 @@ func TestSendHTML(t *testing.T) {
 
 	t.Run("error", func(t *testing.T) {
 		w := httptest.NewRecorder()
-		err := SendHTML(w, nil, struct{}{})
+		err := SendHTML(w, httptest.NewRequest("", "/", nil), struct{}{})
 		require.Error(t, err)
 		require.Equal(t, "text/html; charset=utf-8", w.Header().Get("Content-Type"))
 	})
@@ -481,7 +481,7 @@ func TestSendHTML(t *testing.T) {
 func TestSendHTMLError(t *testing.T) {
 	t.Run("base", func(t *testing.T) {
 		w := httptest.NewRecorder()
-		SendHTMLError(w, nil, errors.New("Hello World"))
+		SendHTMLError(w, httptest.NewRequest("", "/", nil), errors.New("Hello World"))
 
 		require.Equal(t, http.StatusInternalServerError, w.Result().StatusCode)
 		require.Equal(t, "text/html; charset=utf-8", w.Header().Get("Content-Type"))
@@ -489,7 +489,7 @@ func TestSendHTMLError(t *testing.T) {
 	})
 	t.Run("error with status", func(t *testing.T) {
 		w := httptest.NewRecorder()
-		SendHTMLError(w, nil, BadRequestError{Title: "My Error", Detail: "Greeting the world didn't work", Err: errors.New("Hello World failed")})
+		SendHTMLError(w, httptest.NewRequest("", "/", nil), BadRequestError{Title: "My Error", Detail: "Greeting the world didn't work", Err: errors.New("Hello World failed")})
 		require.Equal(t, http.StatusBadRequest, w.Result().StatusCode)
 		require.Equal(t, "text/html; charset=utf-8", w.Header().Get("Content-Type"))
 		require.Equal(t, "400 My Error (Greeting the world didn't work)", w.Body.String())

--- a/tests_test.go
+++ b/tests_test.go
@@ -16,7 +16,7 @@ func TestContentType(t *testing.T) {
 
 	t.Run("Sends application/problem+json when return type is HTTPError", func(t *testing.T) {
 		GetStd(server, "/json-problems", func(w http.ResponseWriter, r *http.Request) {
-			SendJSONError(w, nil, UnauthorizedError{
+			SendJSONError(w, r, UnauthorizedError{
 				Title: "Unauthorized",
 			})
 		})
@@ -32,7 +32,7 @@ func TestContentType(t *testing.T) {
 
 	t.Run("Sends application/json when return type is not HTTPError", func(t *testing.T) {
 		GetStd(server, "/json", func(w http.ResponseWriter, r *http.Request) {
-			SendJSONError(w, nil, errors.New("error"))
+			SendJSONError(w, r, errors.New("error"))
 		})
 
 		req := httptest.NewRequest("GET", "/json", nil)

